src/ansys/acp/core/_recursive_copy.py

```diff
@@ -0,0 +1,246 @@
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from collections.abc import Iterable
+from dataclasses import dataclass
+
+import networkx as nx
+
+from ._tree_objects import LookUpTable1D, LookUpTable1DColumn, LookUpTable3D, LookUpTable3DColumn
+from ._tree_objects._traversal import (
+    all_linked_objects,
+    child_objects,
+    directly_linked_objects,
+    edge_property_lists,
+    edge_property_targets,
+    linked_object_lists,
+)
+from ._tree_objects.base import CreatableTreeObject, TreeObject
+
+__all__ = ["recursive_copy"]
+
+
+@dataclass
+class _WalkTreeOptions:
+    include_children: bool
+    include_linked_objects: bool
+
+
+def _build_dependency_graph(
+    *, source_objects: Iterable[CreatableTreeObject], options: _WalkTreeOptions
+) -> tuple[nx.DiGraph, dict[str, CreatableTreeObject]]:
+    graph = nx.DiGraph()
+    visited_objects: dict[str, CreatableTreeObject] = dict()
+    for tree_object in source_objects:
+        _build_dependency_graph_impl(
+            tree_object=tree_object, graph=graph, visited_objects=visited_objects, options=options
+        )
+    return graph, visited_objects
+
+
+def _build_dependency_graph_impl(
+    *,
+    tree_object: CreatableTreeObject,
+    graph: nx.DiGraph,
+    visited_objects: dict[str, CreatableTreeObject],
+    options: _WalkTreeOptions,
+) -> None:
+    key = tree_object._resource_path.value
+    if key in visited_objects:
+        return
+    visited_objects[key] = tree_object
+    graph.add_node(key)
+
+    if options.include_children:
+        for child_object in child_objects(tree_object):
+            if not isinstance(child_object, CreatableTreeObject):
+                continue
+            graph.add_edge(child_object._resource_path.value, key)
+            _build_dependency_graph_impl(
+                tree_object=child_object,
+                graph=graph,
+                visited_objects=visited_objects,
+                options=options,
+            )
+    if options.include_linked_objects:
+        for linked_object in all_linked_objects(tree_object):
+            assert isinstance(linked_object, CreatableTreeObject)
+            graph.add_edge(tree_object._resource_path.value, linked_object._resource_path.value)
+            _build_dependency_graph_impl(
+                tree_object=linked_object,
+                graph=graph,
+                visited_objects=visited_objects,
+                options=options,
+            )
+
+
+def recursive_copy(
+    *,
+    source_objects: Iterable[CreatableTreeObject],
+    parent_mapping: Iterable[tuple[TreeObject, TreeObject]],
+    include_children: bool = True,
+    include_linked_objects: bool = True,
+) -> list[CreatableTreeObject]:
```

Comment:
Would be better to have a dict mapping old to new objects.

TODO: Investigate whether tree objects can be made hashable (use their resource path or `id()`) - resource path would be "nicer", but maybe problematic since it exists only after storing.

Reply:
done - added a `__hash__` using the `id()`, `__eq__` exists already.